File: clients/java/open-feature-provider/src/main/java/io/superposition/openfeature/EvaluationArgs.java
package io.superposition.openfeature;

import dev.openfeature.sdk.EvaluationContext;
import dev.openfeature.sdk.Value;
import io.juspay.superposition.model.ContextPartial;
import io.juspay.superposition.model.GetConfigOutput;
import com.google.gson.Gson;
import software.amazon.smithy.java.core.serde.document.Document;
import uniffi.superposition_client.MergeStrategy;
import uniffi.superposition_client.OperationException;
import uniffi.superposition_client.Superposition_clientKt;
import uniffi.superposition_types.Context;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

class EvaluationArgs {
    private static final Gson gson = new Gson();

    private static String valueToJsonString(Object value) {
        return gson.toJson(value);
    }

    private static Map<String, String> toQueryData(EvaluationContext eContext) {
        var m = eContext.asObjectMap();
        if (m == null) {
            return new HashMap();
        }
        return m.entrySet().stream()
            .collect(Collectors.toMap(
                Map.Entry::getKey,
                entry -> valueToJsonString(entry.getValue())
            ));
    }

    private static Object documentToObject(Document d) {
        return d.asObject();
    }

    private static String serializeDocument(Document d) {
        return valueToJsonString(documentToObject(d));
    }

    private static Map<String, String> serializeDocumentValues(Map<String, Document> m) {
        return m.entrySet().stream().collect(
            Collectors.toMap(
                Map.Entry::getKey,
                entry -> serializeDocument(entry.getValue())
            )
        );
    }

    private static Context toFfiContext(ContextPartial cp) {
        var cond = serializeDocumentValues(cp.condition());
        return new Context(
            cp.id(),
            cond,
            cp.priority(),
            cp.weight(),
            cp.overrideWithKeys()
        );
    }

    private static Context toFfiContext(SuperpositionConfig.Context ctx) {
        var cond = ctx.condition.entrySet().stream().collect(Collectors.toMap(
            Map.Entry::getKey,
            entry -> valueToJsonString(entry.getValue())
        ));
        return new Context(
            ctx.id,
            cond,
            ctx.priority,
            ctx.weight,
            ctx.overrideWithKeys
        );
    }

    private static Object toSerializable(Value v) {
        if (v.isStructure()) {
            var struct = v.asStructure();
            return struct.asMap().entrySet().stream().collect(Collectors.toMap(
                Map.Entry::getKey,
                entry -> toSerializable(entry.getValue())
            ));
        }

        if (v.isList()) {
            var list = v.asList();
            return list.stream().map(EvaluationArgs::toSerializable).collect(Collectors.toList());
        }

        if (v.isInstant()) {
            // REVIEW This might be incorrect.
            return v.asInstant().toEpochMilli();
        }

        return v.asObject();
    }

    // Values are serialized json values.
    final Map<String, String> defaultConfig;
    final List<Context> contexts;
    // Values in the 2nd Map are serialized json values.
    final Map<String, Map<String, String>> overrides;

    Map<String, String> evaluate(EvaluationContext queryContext) throws OperationException {
        var query = toQueryData(queryContext);
        return Superposition_clientKt.ffiEvalConfig(
            defaultConfig,
            contexts,
            overrides,
            query,
            MergeStrategy.MERGE,
            null
        );
    }

    EvaluationArgs(GetConfigOutput output) {
        defaultConfig = serializeDocumentValues(output.defaultConfigs());
        contexts = output.contexts().stream()
            .map(EvaluationArgs::toFfiContext)
            .collect(Collectors.toList());
        overrides = output.overrides().entrySet().stream().collect(
            Collectors.toMap(
                Map.Entry::getKey,
                entry -> serializeDocumentValues(entry.getValue())
            )
        );
    }

    EvaluationArgs(SuperpositionConfig config) {
        defaultConfig = config.defaultConfig
            .entrySet()
            .stream()
            .collect(Collectors.toMap(
                Map.Entry::getKey,
                entry -> valueToJsonString(toSerializable(entry.getValue()))));
        contexts = config.contexts.stream()
            .map(EvaluationArgs::toFfiContext)
            .collect(Collectors.toList());
        overrides = config.overrides.entrySet().stream().collect(Collectors.toMap(
            Map.Entry::getKey,
            entry -> entry.getValue().entrySet().stream().collect(Collectors.toMap(
                Map.Entry::getKey,
                _entry -> valueToJsonString(toSerializable(_entry.getValue()))
            ))
        ));
    }
}

File: clients/java/openfeature-provider/src/main/kotlin/io/juspay/superposition/openfeature/EvaluationArgs.kt
package io.juspay.superposition.openfeature

import com.google.gson.Gson
import dev.openfeature.sdk.EvaluationContext
import dev.openfeature.sdk.Value
import io.juspay.superposition.model.ContextPartial
import io.juspay.superposition.model.ExperimentResponse
import io.juspay.superposition.model.ExperimentGroupResponse
import io.juspay.superposition.model.GetConfigOutput
import software.amazon.smithy.java.core.serde.document.Document
import uniffi.superposition_client.*
import uniffi.superposition_types.Buckets
import uniffi.superposition_types.Context
import uniffi.superposition_types.DimensionInfo
import uniffi.superposition_types.DimensionType
import uniffi.superposition_types.GroupType
import uniffi.superposition_types.Variant
import uniffi.superposition_types.VariantType

internal class EvaluationArgs {
    // Values are serialized json values.
    val defaultConfig: Map<String, String>
    val contexts: List<Context>
    // Values in the 2nd Map are serialized json values.
    val overrides: Map<String, Map<String, String>>
    val dimensions: Map<String, DimensionInfo>

    @Throws(OperationException::class)
    fun evaluate(queryContext: EvaluationContext, eargs: ExperimentationArgs?): MutableMap<String, String> {
        val query = toQueryData(queryContext)
        return ffiEvalConfig(
            defaultConfig,
            contexts,
            overrides,
            dimensions,
            query,
            MergeStrategy.MERGE,
            null,
            eargs
        ).toMutableMap()
    }

    @Throws(OperationException::class)
    fun getApplicableVariants(ectx: EvaluationContext, experimentationArgs: ExperimentationArgs): List<String> {
        val qdata = toQueryData(ectx)
        return ffiGetApplicableVariants(experimentationArgs, dimensions, qdata, null)
    }

    constructor(output: GetConfigOutput) {
        defaultConfig = serializeDocumentValues(output.defaultConfigs())
        contexts = output.contexts().map { toFfiContext(it) }
        overrides = output.overrides().mapValues { serializeDocumentValues(it.value) }
        dimensions = output.dimensions().mapValues { toFfiDimensionInfo(it.value) }
    }

    constructor(config: SuperpositionConfig) {
        defaultConfig = config.defaultConfig.mapValues { valueToJsonString(toSerializable(it.value)) }
        contexts = config.contexts
        overrides = config.overrides.mapValues {
            it.value.mapValues { e -> valueToJsonString(toSerializable(e.value)) }
        }
        dimensions = config.dimensions
    }

    internal object Helpers {
        private fun toFfiVariant(v: io.juspay.superposition.model.Variant): Variant {
            var vtype = VariantType.CONTROL
            if (v.variantType().type() == io.juspay.superposition.model.VariantType.Type.EXPERIMENTAL) {
                vtype = VariantType.EXPERIMENTAL
            }
            return Variant(
                v.id(),
                vtype,
                v.contextId(),
                v.overrideId(),
                serializeDocumentValues(v.overrides().asStringMap())
            )
        }

        @JvmStatic
        fun toFfiExperiment(er: ExperimentResponse): FfiExperiment {
            val variants = er.variants().map { toFfiVariant(it) }
            return FfiExperiment(
                er.id(),
                er.trafficPercentage().toUByte(),
                variants,
                serializeDocumentValues(er.context())
            )
        }

        private fun toFfiGroupType(gt: io.juspay.superposition.model.GroupType): GroupType {
            return when (gt) {
                io.juspay.superposition.model.GroupType.USER_CREATED -> GroupType.USER_CREATED
                else -> GroupType.SYSTEM_GENERATED
            }
        }

        @JvmStatic
        fun toFfiExperimentGroup(er: ExperimentGroupResponse): FfiExperimentGroup {
            return FfiExperimentGroup(
                er.id(),
                serializeDocumentValues(er.context()),
                er.trafficPercentage().toUByte(),
                er.memberExperimentIds(),
                toFfiGroupType(er.groupType()),
                er.buckets() as Buckets
            )
        }
    }

    companion object {
        private val gson = Gson()

        private fun valueToJsonString(value: Any): String {
            return gson.toJson(value)
        }

        private fun toQueryData(eContext: EvaluationContext): Map<String, String> {
            val m = eContext.asObjectMap()
            if (m == null) {
                return HashMap()
            }
            return m.mapValues { valueToJsonString(it.value) }
        }

        private fun serializeDocument(d: Document): String {
            return valueToJsonString(d.asObject())
        }

        private fun serializeDocumentValues(m: MutableMap<String, Document>): Map<String, String> {
            return m.mapValues { serializeDocument(it.value) }
        }

        private fun toFfiContext(cp: ContextPartial): Context {
            val cond = serializeDocumentValues(cp.condition())
            return Context(
                cp.id(),
                cond,
                cp.priority(),
                cp.weight(),
                cp.overrideWithKeys()
            )
        }

        private fun toFfiDimensionInfo(dim: io.juspay.superposition.model.DimensionInfo): DimensionInfo {
            return DimensionInfo(
                serializeDocumentValues( dim.schemaMember() ),
                dim.position(),
                toFfiDimensionType (dim.dimensionType() ),
                dim.dependencyGraph()
            )
        }

        private fun toFfiDimensionType(dimType: io.juspay.superposition.model.DimensionType): DimensionType {
            return when (dimType.type()) {
                io.juspay.superposition.model.DimensionType.Type.localCohort -> DimensionType.LocalCohort(dimType.getValue<String>())
                io.juspay.superposition.model.DimensionType.Type.remoteCohort -> DimensionType.RemoteCohort(dimType.getValue<String>())
                else -> DimensionType.Regular
            }
        }

        private fun toSerializable(v: Value): Any {
            if (v.isStructure) {
                val struct = v.asStructure()
                return struct.asMap().mapValues { toSerializable(it.value) }
            }

            if (v.isList) {
                return v.asList().map { toSerializable(it) }
            }

            if (v.isInstant) {
                return v.asInstant().toEpochMilli()
            }

            return v.asObject()
        }
    }
}
